src/lib/ngsiNotify/QueueWorkers.cpp

```diff
@@ -247,6 +260,25 @@ static void* workerFunc(void* pSyncQ)
           QueueStatistics::incSentError();
           alarmMgr.notificationError(url, "notification failure for queue worker: " + out);
 
+          params->failsCounter++;
+
+          if ((params->failsCounter)  > (params->maxFailsLimit))
```

Comment:
Simplify to:

```suggestion
          if (params->failsCounter > params->maxFailsLimit)
```

In addition, please review indent of the if block. [Style rule](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/contribution_guidelines.md#m4-indent) is to use 2 spaces, not 1 (see )

Reply:
In fact, there are bigger problems in this part:

* Modifications done to the params (such as `params->failsCounter = 0` or `params->failsCounter++`) are not going to be persisted in the subscription (either cache nor DB). Note that `params` is a read-only object.
* You should avoid direct access to cache/DB in this `workerFunc()`. That access should be done in the `subNotificationErrorStatus()` function, which is though to update subscription information when the notification fail, both in cache and DB. I think you have done some modifications to that function in this PR but maybe not all them (in particular, the check to pass the status to inactive)

Reply:
> In fact, there are bigger problems in this part:
> 
> * Modifications done to the params (such as `params->failsCounter = 0` or `params->failsCounter++`) are not going to be persisted in the subscription (either cache nor DB). Note that `params` is a read-only object.
> * You should avoid direct access to cache/DB in this `workerFunc()`. That access should be done in the `subNotificationErrorStatus()` function, which is though to update subscription information when the notification fail, both in cache and DB. I think you have done some modifications to that function in this PR but maybe not all them (in particular, the check to pass the status to inactive)

I didn't understand clearly what change we need to do in `src/lib/ngsiNotify/QueueWorkers.cpp` and `src/lib/ngsiNotify/senderThread.cpp` .

We need to add params->failsCounter in subNotificationErrorStatus() instread of using directly in `src/lib/ngsiNotify/QueueWorkers.cpp` and `src/lib/ngsiNotify/senderThread.cpp`? Also please confirm the code of setting  status=inactive in DB and cache after comparison of `if (params->failsCounter > params->maxFailsLimit)` is correct or we need to change this code also?

Also, Currently the default value of params->failsCounter is long long random number instead of 0. Can you suggest where we should set the value of this params->failsCounter.

Reply:
The bottom line (following the principle of "separation of concerns" explained also [here](https://github.com/telefonicaid/fiware-orion/pull/3890#discussion_r677511495)) is that the code in `workerFunc()` and `startSenderThread()` only deals with the notification sending and detecting error but the process done in case of error (including updating subscription in cache and DB) is delegated to `subNotificationErrorStatus()`). You shouldn't need any of this in this file:

```
#include "cache/subCache.h"
#include "mongoDriver/connectionOperations.h"

#include "mongoBackend/MongoGlobal.h"
#include "mongoDriver/safeMongo.h"
#include "mongoBackend/MongoCommonUpdate.h"
#include "mongoBackend/dbConstants.h"
#include "ngsi/MaxFailsLimit.h"
```

The current version of `subNotificationErrorStatus()` is fine, but I think it should also include the subscription state change in the case of reaching maximum number of allowed fails ("moving" the block in L265-L280 currently in src/lib/ngsiNotify/QueueWorkers.cpp to `subNotificationErrorStatus()`).

With regards to default value for `params->failsCounter` I understand the place to initialize it is the one in which you are setting all the other fields in `params`, i.e. the `buildSenderParams()` function (as explained [here](https://github.com/telefonicaid/fiware-orion/pull/3890#discussion_r677511495)). It should be initialized with the values coming from the subscription,

Reply:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3890/commits/6b68a8c9ded45dd2e30c033f4d5000cddfab19de